### ISSUE TITLE
System.IO.Compression: Path validation of zip entries is done based on current OS

### ISSUE THREAD

**Issue by ianhays** [dotnet/corefx#4991]
When reading a Zip file, calls are made to Path validation that check invalid characters based on the current OS. This brings about ArgumentException errors with the message "Illegal characters in path."

Example repro:
- Linux commands:
  - touch "ab\<cd>"
  - zip newzip.zip "ab\<cd>"
- copy newzip.zip to Windows
- Windows console app:

```
using System;
using System.Diagnostics;
using System.IO;
using System.IO.Compression;
class Program
{
    static void Main()
    {            
        using (ZipArchive archive = new ZipArchive(new FileStream("newzip.zip", FileMode.Open), ZipArchiveMode.Read))
            {
                foreach (var entry in archive.Entries)
                {

                }
            }
            
    }
}
```

Result:
```
Unhandled Exception: System.ArgumentException: Illegal characters in path.
arameter name: path
  at System.IO.PathInternal.CheckInvalidPathChars(String path, Boolean checkAdditional)
  at System.IO.Path.GetFileName(String path)
  at System.IO.Compression.ZipHelper.EndsWithDirChar(String test)
  at System.IO.Compression.ZipArchiveEntry.set_FullName(String value)
  at System.IO.Compression.ZipArchiveEntry..ctor(ZipArchive archive, ZipCentralDirectoryFileHeader cd)
  at System.IO.Compression.ZipArchive.ReadCentralDirectory()
  at System.IO.Compression.ZipArchive.get_Entries()
  at Program.Main()
```

We should at least throw a more useful error message like "Zip file contains files with invalid characters in the path names". We should also consider modifying the checks to do path validation based on the zip's source OS instead of the current OS but will need to be careful to think about the ramifications of returning paths from Name and FullName that are invalid on the current system.

@stephentoub @justinvp

**Comment by justinvp** [comment]
I wonder how other Windows unzip utilities/APIs (e.g. Explorer.exe, 7zip, WinZip, java's zip API, etc.) behave when unzipping newzip.zip with Linux paths?

**Comment by ianhays** [comment]
Backslashes in the file name are interpreted as file separators by Explorer and 7Zip regardless of source OS.

files with invalid chars in the name are ignored by the Explorer window. The zip shows as empty. When attempting to extract I get the error message "Windows cannot complete the extraction. The compressed zipped folder is invalid".

With 7zip the file shows up with the correct name in the window. Extracting it yields the file "ab_c_d", so it replaced the invalid characters with underscores.

**Comment by ianhays** [comment]
info-zip on Windows does the same thing that 7Zip does.

**Comment by justinvp** [comment]
It's interesting that the repro above throws... Should it? I can imagine scenarios where someone wants to read all entries from zip, being able to see the values of `FullName` (and maybe even `Name`) without these properties throwing, regardless of whether they have invalid chars or mismatched dir separators from the current OS.

Perhaps only the file-system-based extensions in `System.IO.Compression.ZipFile` should throw when extracting to a directory or file?

This way, if someone wanted to use these APIs to create an extraction utility that replaced invalid chars with underscores or whatever, they could read the entries (without throwing) and do the actual extraction to the file system themselves where they handle replacing the chars, instead of relying on the `ZipFile` helpers, which would throw.

**Comment by ianhays** [comment]
>It's interesting that the repro above throws... Should it? 

I do not think that we are throwing at the correct time. I agree with you that FullName and Name should be returning exactly the Zip intended for them to be i.e. they shouldn't be windows-ized paths if it is a Unix zip.

Doing path validity checks on the entries within a ZipArchive does make sense as it serves as a check for the validity of the entire ZipArchive, but only if we're doing the <i>right</i> checks. What we're doing now is saying that a zip with paths invalid on this machine is an invalid zip, but that isn't necessarily true.

>Perhaps only the file-system-based extensions in System.IO.Compression.ZipFile should throw when extracting to a directory or file?

That's what I was thinking, yeah. In ZipFile Extraction we should either throw like Explorer or replace the invalid chars with underscores like info-zip. The current implementation only uses the FullName of a ZipArchiveEntry, so backslashes are treated as separators and "<>" etc. are treated as invalid.

**Comment by ianhays** [comment]
@stephentoub @mellinoe any thoughts?

**Comment by stephentoub** [comment]
I agree we shouldn't be throwing in this example.  We should provide full-fidelity (or best possible fidelity) regardless of OS, and only use a fallback scheme when attempting to extract the archive, at which point we do need to handle any mismatches between what's in the archive and the rules of the target file system.  For an invalid character, the options would seem to include throwing, providing a substitute character that is valid, or ignoring the entry.  If we were to throw, is there a workaround that would let the developer still extract the data through other APIs?

**Comment by ianhays** [comment]
>For an invalid character, the options would seem to include throwing, providing a substitute character that is valid, or ignoring the entry.

Yes. The same options exist for how to treat the backslash on a Unix-Compiled Zip.

>If we were to throw, is there a workaround that would let the developer still extract the data through other APIs?

The only places I can think of that throwing potentially makes sense would be:
- In the Name getter of ZipArchiveEntry. If we throw here then there isn't any other API to get the path beyond manually parsing the stream. 
- When trying to use the ZipFile Extraction extension methods.Throwing here is less of an issue because we could always manually open the ZipArchive and recreate it with valid paths, but that would be a nontrivial amount of work.

**Comment by stephentoub** [comment]
> The only places I can think of that throwing potentially makes sense would be:

Sorry if I wasn't clear.  I do not believe we shoudl throw anywhere other than in APIs whose sole purpose is to write to the file system, thereby needing to use a valid name, and thereby running into problems if the name in the zip is invalid for the target file system.  Do you have a list of the APIs that would be affected then?  If such APIs have workarounds such that, if they would throw, there are other APIs that could be used to achieve the same behavior, then throwing from them might not be bad, but if there isn't a workaround / no other APIs that could be used, then we shouldn't throw and should instead choose a fall back like character substitution.

**Comment by ianhays** [comment]
I think we're on the same page; you want to know which Extraction methods would be broken by using the 'true' paths in ZipArchive and how difficult it is to work around.

```ZipArchive.ExtractToFile``` would be fine since it takes a path as a parameter and doesn't use the entry's path.

```ZipArchive.ExtractToDirectory``` would throw since it attempts to create files from ArchiveEntries using the path in the entry. A workaround is manually iterating through the entries in the Archive and doing your own substitution on the names.

To sum up, we're now discussing a behavior like so for zips that have invalid paths:
- Reading a Zip :
  - The ZipArchiveEntry.Name property returns the correct filename of the entry based on the source OS, including backslashes or invalid chars (if applicable)
-  Extracting a Zip 
  - ExtractToDirectory throws ```System.ArgumentException: Illegal characters in path.```.

**Comment by stephentoub** [comment]
> I think we're on the same page

Yup.  This sounds good to me.  Thanks.

**Comment by henning-krause** [comment]
Will this fix eventually be backported to .NET Framework 4.x? .NET up to 4.7 suffers from the same error.
